tests/test_verify.py

```diff
@@ -0,0 +1,110 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Test verification."""
+# pylint: disable=unused-argument
+
+import os
+import copy
+import pytest
+
+import torch
+from torch import nn
+import torch.nn.functional as F
+
+import slapo
+from slapo.pattern import call_module
+
+
+def test_vertical_fusion():
+    class Model(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.conv = nn.Conv2d(3, 32, 3)
+
+        def forward(self, x):
+            x = self.conv(x)
+            x = F.relu(x)
+            x = x + 1
+            return x
+
+    mod = Model().cuda()
+    sch = slapo.create_schedule(mod)
+
+    def pattern(x: torch.Tensor):
+        x = call_module("conv", x)
+        x = F.relu(x)
+        x = x + 1
+        return x
+
+    subgraph = sch.find(pattern)
+    assert len(subgraph[0]) == 3
+    inp = torch.randn((1, 3, 32, 32), requires_grad=True).cuda()
+    with slapo.verify(sch, inp):
+        sch.fuse(subgraph, compiler="TorchScript", name="FusedReLU")
+
+
+def test_bias_gelu():
+    class Model(nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear = nn.Linear(1024, 1024)
+            self.gelu = nn.GELU()
+
+        def forward(self, x):
+            x = self.linear(x)
+            x = self.gelu(x)
+            return x
+
+    mod = Model().cuda()
+    sch = slapo.create_schedule(mod)
+
+    sch["linear"].decompose()
+    sch.trace(flatten=True)
+
+    def pattern(x, bias):
+        x = F.gelu(bias + x)
+        return x
+
+    subgraph = sch.find(pattern)
+    assert len(subgraph[0]) == 2
+    inp = torch.randn((1, 16, 1024, 1024), requires_grad=True)
+    with slapo.verify(sch, inp):
+        sch.fuse(subgraph, compiler="TorchScript", name="BiasGeLU")
+    assert isinstance(sch["BiasGeLU_0"].mod, torch.jit.ScriptModule)
+
+
+def test_linear(init_dist):
+    class Model(torch.nn.Module):
+        def __init__(self):
+            super().__init__()
+            self.linear1 = torch.nn.Linear(20, 30)
+            self.linear2 = torch.nn.Linear(30, 40)
+
+        def forward(self, data):
+            out = self.linear1(data)
+            out = self.linear2(out)
+            return out
+
+    local_rank = int(os.environ["LOCAL_RANK"])
+    torch.cuda.set_device(local_rank)
+    model = Model()
+
+    sch = slapo.create_schedule(copy.deepcopy(model))
+    with slapo.verify(sch, [torch.randn(10, 20)], device=f"cuda:{local_rank}"):
+        sch["linear1"].shard("weight", axis=0)
+        sch["linear1"].shard("bias", axis=0)
+        sch["linear1"].sync(mode="bwd_post", sync_op_or_fn="all_reduce")
+        sch["linear2"].shard("weight", axis=1)
```

Comment:
Is it possible to support verify part of the schedule? e.g., just verify the second layer sharding.

Reply:
But how can we verify the partial results? We can for example only verify the first layer sharding, but this requires an explicit synchronization point (all_gather) after the first linear layer.

Reply:
Yes, then we should be able to insert an assertion point after all_gather, if we need/want to.

Reply:
The point is if we automatically insert the synchronization then there will be no differences between the following two cases:
```python
with slapo.verify(sch, [torch.randn(10, 20)], device=f"cuda:{local_rank}"):
    sch["linear1"].shard("weight", axis=0)
    sch["linear1"].shard("bias", axis=0)
```
and
```python
with slapo.verify(sch, [torch.randn(10, 20)], device=f"cuda:{local_rank}"):
    sch["linear1"].shard("weight", axis=0)
    sch["linear1"].shard("bias", axis=0)
    sch["linear1"].sync(mode="bwd_post", sync_op_or_fn="all_gather")
```
which means users do not know whether they miss inserting a synchronization primitive.

Reply:
At least the primitives inside the `verify` context should be self-contained, meaning after applying this sequence of primitives, the outputs should be identical to the original one.

Reply:
No, we don't need to insert the sync point automatically. The point is that we need the ability to break a large schedule piece by piece.

Reply:
If you mean verify a part of the schedule, then yes, we can definitely do it, as long as that part of the schedule is self-contained, as what I shown in the second example.

Reply:
IMHO we could allow users to provide a post-process function for verification. For example, in this case, the post-processing function should have an all-gather.

Reply:
> For example, in this case, the post-processing function should have an all-gather.

Then why not directly let user write a `.sync(all_gather)` primitive?

> Another idea is allowing users to custom the verify function

This probably makes things complicated for now. I think the existing interface has already placed some burden on programmers, as they are required to refactor the schedule code in order to leverage the verifier. If the verification process also needs to be customized, the schedule code will become cumbersome.

Reply:
For example when we want to verify an attention block with sequence parallel, all-gather should not be in the schedule. Just like our sharding unit tests, which we manually all-gather outputs and gradients for verification.